pkg/server/plugin/nodeattestor/awsiid/eks.go

```diff
@@ -0,0 +1,315 @@
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	// Make sure: we are not doing cache burst and account map is not updated recently from different go routine.
+	if !catchBurst && len(o.eksNodeList) != 0 && !o.checkIfTTLIsExpired(o.eksNodeListValidDuration) {
+		return o.eksNodeList, nil
+	}
+
+	// Avoid if other thread has already updated the map
+	if catchBurst && o.retries == 0 {
+		return o.eksNodeList, nil
+	}
+
+	// Get the list of node groups belonging to the EKS cluster
+	listNodegroupsOp, err := eksClient.ListNodegroups(ctx, &eks.ListNodegroupsInput{
+		ClusterName: &o.eksConfig.EKSClusterName,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("issue while getting list of EKS Nodegroups: %w", err)
+	}
+
+	// Build new EKS nodes list
+	eksNodeMap := make(map[string]any)
+
+	for {
+		for _, ng := range listNodegroupsOp.Nodegroups {
+			instances, err := o.fetchNodesInNodeGroup(ctx, eksClient, asClient, ng)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, instance := range instances {
+				eksNodeMap[instance] = struct{}{}
+			}
+		}
+
+		if listNodegroupsOp.NextToken == nil {
+			break
+		}
+
+		listNodegroupsOp, err = eksClient.ListNodegroups(ctx, &eks.ListNodegroupsInput{
+			ClusterName: &o.eksConfig.EKSClusterName,
+			NextToken:   listNodegroupsOp.NextToken,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("issue while getting list of EKS Nodegroups in pagination: %w", err)
+		}
+	}
+
+	// Update timestamp, if it was not invoked as part of cache miss.
+	if !catchBurst {
+		o.eksNodeListValidDuration = o.clk.Now().UTC().Add(o.eksAccountListCacheTTL)
+		// Also reset the retries
+		o.retries = orgAccountRetries
+	}
+
+	// Overwrite the cache/list
+	o.eksNodeList = eksNodeMap
+
+	return o.eksNodeList, nil
+}
+
+// reloadNodeList gets the list of nodes belonging to the EKS cluster and catch them
+func (o *eksValidator) fetchNodesInNodeGroup(ctx context.Context, eksClient EKSClient, asClient autoscaling.DescribeAutoScalingGroupsAPIClient, nodeGroup string) ([]string, error) {
+	// Get the list of node groups belonging to the EKS cluster
+	describeNodegroupOp, err := eksClient.DescribeNodegroup(ctx, &eks.DescribeNodegroupInput{
+		ClusterName:   &o.eksConfig.EKSClusterName,
+		NodegroupName: &nodeGroup,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("issue while getting list of EKS Node Groups: %w", err)
+	}
+
+	instances := make([]string, 0)
+	for _, asg := range describeNodegroupOp.Nodegroup.Resources.AutoScalingGroups {
+		if asg.Name == nil {
+			continue
+		}
+
+		// Get the list of instances in the AutoScalingGroup
+		describeASGOp, err := asClient.DescribeAutoScalingGroups(ctx, &autoscaling.DescribeAutoScalingGroupsInput{
+			AutoScalingGroupNames: []string{*asg.Name},
+		})
+
+		if err != nil {
+			return nil, fmt.Errorf("issue while getting list of instances in AutoScalingGroup: %w", err)
+		}
+
+		for {
+			for _, instance := range describeASGOp.AutoScalingGroups[0].Instances {
```

Comment:
Only examining the first autoscaling group looks like a bug to me. Is there any reason for limiting to only the first autoscaling group in the node group?

Reply:
It was one of those quick PoC code bugs :smile: Will fix it in the next commit